CLIENT-3439 Provide user-level metrics for applications using Aerospike Database (Metrics v2)

mirzakaracic opened the pull request:
Adding detailed metrics and resultcodes with namespaces to golang client.

- [x] Added latency for node/namespace/ and cluster 
- [x] Added bytes-sent bytes-received for node/namespace/command and cluster 
- [x] Added connection acquired for node/namespace/command and cluster 
- [x] Added parsing latency for node/namespace/command and cluster 
- [x] Added all (0 - 255) server error counts for node/namespace/command and cluster 
- [x] Added labels for the cluster

Valided and optimized updateOrInsert calls. Below is a list of benchmarks as result of those optimizations.

```go
	_, err = func() (int, Error) {
		if metricsEnabled {
			var dataSent = 0
			start := time.Now()
			dataSent, err = cmd.conn.Write(cmd.dataBuffer[:cmd.dataOffset])
			cmd.applyDetailedMetricsBytesSentAndTransmission(ifc, dataSent, start)
			return dataSent, err
		}
		return cmd.conn.Write(cmd.dataBuffer[:cmd.dataOffset])
	}()
```

=== RUN   BenchmarkApplyDetailedMetricsBytesSentAndTransmission
BenchmarkApplyDetailedMetricsBytesSentAndTransmission
BenchmarkApplyDetailedMetricsBytesSentAndTransmission-14         8375888               131.0 ns/op            24 B/op          1 allocs/op
=== RUN   BenchmarkApplyDetailedConnectionAq
BenchmarkApplyDetailedConnectionAq
BenchmarkApplyDetailedConnectionAq-14                           11469990               103.1 ns/op            24 B/op          1 allocs/op
=== RUN   BenchmarkApplyDetailedParsing
BenchmarkApplyDetailedParsing
BenchmarkApplyDetailedParsing-14                                11457337               104.4 ns/op            24 B/op          1 allocs/op
=== RUN   BenchmarkCommandMergeCommandResultCodeMetrics
BenchmarkCommandMergeCommandResultCodeMetrics
BenchmarkCommandMergeCommandResultCodeMetrics-14                 3045164               401.4 ns/op            64 B/op          5 allocs/op
=== RUN   BenchmarkCommandMergeDetailMetrics
BenchmarkCommandMergeDetailMetrics
BenchmarkCommandMergeDetailMetrics-14                            5773930               202.5 ns/op            24 B/op          2 allocs/op
=== RUN   BenchmarkConnectionAcWithMetrics
BenchmarkConnectionAcWithMetrics
BenchmarkConnectionAcWithMetrics-14                             10457097               115.9 ns/op            24 B/op          1 allocs/op
=== RUN   BenchmarkParseAcWithMetrics
BenchmarkParseAcWithMetrics
BenchmarkParseAcWithMetrics-14                                  10541271               114.7 ns/op            24 B/op          1 allocs/op
PASS


Benchmark results without  inline optimizations, aka  on metrics callection in the command.executeAt method. As example
 
```go
	if metricsEnabled {
		start := time.Now()
		dataSent, err = cmd.conn.Write(cmd.dataBuffer[:cmd.dataOffset])
		cmd.applyDetailedMetricsBytesSentAndTransmission(ifc, dataSent, start)
	} else {
		_, err = cmd.conn.Write(cmd.dataBuffer[:cmd.dataOffset])
	}
```

=== RUN   BenchmarkApplyDetailedMetricsBytesSentAndTransmission
BenchmarkApplyDetailedMetricsBytesSentAndTransmission
BenchmarkApplyDetailedMetricsBytesSentAndTransmission-14         8488948               132.0 ns/op            24 B/op          1 allocs/op
=== RUN   BenchmarkApplyDetailedConnectionAq
BenchmarkApplyDetailedConnectionAq
BenchmarkApplyDetailedConnectionAq-14                           11312936               107.7 ns/op            24 B/op          1 allocs/op
=== RUN   BenchmarkApplyDetailedParsing
BenchmarkApplyDetailedParsing
BenchmarkApplyDetailedParsing-14                                10988397               108.8 ns/op            24 B/op          1 allocs/op
=== RUN   BenchmarkCommandMergeCommandResultCodeMetrics
BenchmarkCommandMergeCommandResultCodeMetrics
BenchmarkCommandMergeCommandResultCodeMetrics-14                 2924882               412.2 ns/op            64 B/op          5 allocs/op
=== RUN   BenchmarkCommandMergeDetailMetrics
BenchmarkCommandMergeDetailMetrics
BenchmarkCommandMergeDetailMetrics-14                            5800849               207.9 ns/op            24 B/op          2 allocs/op


------------> benchmarks without inline optmizations

=== RUN   BenchmarkConnectionAcWithMetrics
BenchmarkConnectionAcWithMetrics
BenchmarkConnectionAcWithMetrics-14                              9772182               120.3 ns/op            24 B/op          1 allocs/op
=== RUN   BenchmarkParseAcWithMetrics
BenchmarkParseAcWithMetrics
BenchmarkParseAcWithMetrics-14                                  10514532               115.9 ns/op            24 B/op          1 allocs/op
PASS
ok      github.com/aerospike/aerospike-client-go/v8     9.834s


Filtered result snipped of Stats() snapshot:

```json
{
  "127.0.0.1:3100": {
    "batch-read-metrics": {
      "buckets": [
        0,
        0,
        0,
        3
      ],
      "count": 3,
      "max": 538,
      "min": 451,
      "sum": 1494
    },
    "batch-write-metrics": {
      "buckets": [
        0,
        0,
        0,
        1
      ],
      "count": 1,
      "max": 502,
      "min": 502,
      "sum": 502
    },
    "circuit-breaker-hits": 0,
    "closed-connections": 0,
    "connections-attempts": 100,
    "connections-error-other": 0,
    "connections-error-timeout": 0,
    "connections-failed": 0,
    "connections-idle-dropped": 0,
    "connections-pool-empty": 0,
    "connections-pool-overflow": 0,
    "connections-successful": 100,
    "delete-metrics": {
      "buckets": [
        0,
        0,
        0,
        0
      ],
      "count": 0,
      "max": 0,
      "min": 0,
      "sum": 0
    },
    "detailed-metrics": {
      "test": {
        "BatchRead": {
          "bytes-received": {
            "buckets": [
              0,
              0,
              0,
              3
            ],
            "count": 3,
            "max": 430,
            "min": 214,
            "sum": 858
          },
          "bytes-sent": {
            "buckets": [
              0,
              0,
              0,
              3
            ],
            "count": 3,
            "max": 285,
            "min": 269,
            "sum": 823
          },
          "connection-aq": {
            "buckets": [
              2,
              0,
              0,
              1
            ],
            "count": 3,
            "max": 9,
            "min": 0,
            "sum": 9
          },
          "latency": {
            "buckets": [
              0,
              0,
              1,
              2
            ],
            "count": 3,
            "max": 10,
            "min": 5,
            "sum": 23
          },
          "parsing": {
            "buckets": [
              0,
              0,
              0,
              3
            ],
            "count": 3,
            "max": 511,
            "min": 438,
            "sum": 1444
          }
        },
        "BatchWrite": {
          "bytes-received": {
            "buckets": [
              0,
              0,
              0,
              1
            ],
            "count": 1,
            "max": 214,
            "min": 214,
            "sum": 214
          },
          "bytes-sent": {
            "buckets": [
              0,
              0,
              0,
              1
            ],
            "count": 1,
            "max": 271,
            "min": 271,
            "sum": 271
          },
          "connection-aq": {
            "buckets": [
              1,
              0,
              0,
              0
            ],
            "count": 1,
            "max": 0,
            "min": 0,
            "sum": 0
          },
          "latency": {
            "buckets": [
              0,
              0,
              1,
              0
            ],
            "count": 1,
            "max": 6,
            "min": 6,
            "sum": 6
          },
          "parsing": {
            "buckets": [
              0,
              0,
              0,
              1
            ],
            "count": 1,
            "max": 491,
            "min": 491,
            "sum": 491
          }
        },
        "Put": {
          "bytes-received": {
            "buckets": [
              0,
              0,
              0,
              8
            ],
            "count": 8,
            "max": 30,
            "min": 30,
            "sum": 240
          },
          "bytes-sent": {
            "buckets": [
              0,
              0,
              0,
              8
            ],
            "count": 8,
            "max": 100,
            "min": 100,
            "sum": 800
          },
          "connection-aq": {
            "buckets": [
              8,
              0,
              0,
              0
            ],
            "count": 8,
            "max": 1,
            "min": 0,
            "sum": 1
          },
          "latency": {
            "buckets": [
              0,
              0,
              3,
              5
            ],
            "count": 8,
            "max": 15,
            "min": 5,
            "sum": 64
          },
          "parsing": {
            "buckets": [
              0,
              0,
              0,
              8
            ],
            "count": 8,
            "max": 1263,
            "min": 291,
            "sum": 3575
          }
        }
      }
    },
    "detailed-resultcode-counts": {
      "test": {
        "BatchRead": {
          "OK": 24
        },
        "BatchWrite": {
          "OK": 8
        },
        "Put": {
          "OK": 8
        }
      }
    },
    "exists-metrics": {
      "buckets": [
        0,
        0,
        0,
        0
      ],
      "count": 0,
      "max": 0,
      "min": 0,
      "sum": 0
    },
    "get-header-metrics": {
      "buckets": [
        0,
        0,
        0,
        0
      ],
      "count": 0,
      "max": 0,
      "min": 0,
      "sum": 0
    },
    "get-metrics": {
      "buckets": [
        0,
        0,
        0,
        0
      ],
      "count": 0,
      "max": 0,
      "min": 0,
      "sum": 0
    },
    "node-added-count": 1,
    "node-removed-count": 0,
    "open-connections": 100,
    "operate-metrics": {
      "buckets": [
        0,
        0,
        0,
        0
      ],
      "count": 0,
      "max": 0,
      "min": 0,
      "sum": 0
    },
    "partition-map-updates": 1,
    "put-metrics": {
      "buckets": [
        0,
        0,
        0,
        8
      ],
      "count": 8,
      "max": 1358,
      "min": 302,
      "sum": 3743
    },
    "query-metrics": {
      "buckets": [
        0,
        0,
        0,
        0
      ],
      "count": 0,
      "max": 0,
      "min": 0,
      "sum": 0
    },
    "scan-metrics": {
      "buckets": [
        0,
        0,
        0,
        0
      ],
      "count": 0,
      "max": 0,
      "min": 0,
      "sum": 0
    },
    "tends-failed": 0,
    "tends-successful": 2,
    "tends-total": 2,
    "transaction-error-count": 0,
    "transaction-retry-count": 0,
    "udf-metrics": {
      "buckets": [
        0,
        0,
        0,
        0
      ],
      "count": 0,
      "max": 0,
      "min": 0,
      "sum": 0
    }
  },
  "cluster-aggregated-stats": {
    "batch-read-metrics": {
      "buckets": [
        0,
        0,
        0,
        3
      ],
      "count": 3,
      "max": 538,
      "min": 451,
      "sum": 1494
    },
    "batch-write-metrics": {
      "buckets": [
        0,
        0,
        0,
        1
      ],
      "count": 1,
      "max": 502,
      "min": 502,
      "sum": 502
    },
    "circuit-breaker-hits": 0,
    "closed-connections": 0,
    "connections-attempts": 100,
    "connections-error-other": 0,
    "connections-error-timeout": 0,
    "connections-failed": 0,
    "connections-idle-dropped": 0,
    "connections-pool-empty": 0,
    "connections-pool-overflow": 0,
    "connections-successful": 100,
    "delete-metrics": {
      "buckets": [
        0,
        0,
        0,
        0
      ],
      "count": 0,
      "max": 0,
      "min": 0,
      "sum": 0
    },
    "detailed-metrics": {
      "test": {
        "BatchRead": {
          "bytes-received": {
            "buckets": [
              0,
              0,
              0,
              3
            ],
            "count": 3,
            "max": 430,
            "min": 214,
            "sum": 858
          },
          "bytes-sent": {
            "buckets": [
              0,
              0,
              0,
              3
            ],
            "count": 3,
            "max": 285,
            "min": 269,
            "sum": 823
          },
          "connection-aq": {
            "buckets": [
              2,
              0,
              0,
              1
            ],
            "count": 3,
            "max": 9,
            "min": 0,
            "sum": 9
          },
          "latency": {
            "buckets": [
              0,
              0,
              1,
              2
            ],
            "count": 3,
            "max": 10,
            "min": 5,
            "sum": 23
          },
          "parsing": {
            "buckets": [
              0,
              0,
              0,
              3
            ],
            "count": 3,
            "max": 511,
            "min": 438,
            "sum": 1444
          }
        },
        "BatchWrite": {
          "bytes-received": {
            "buckets": [
              0,
              0,
              0,
              1
            ],
            "count": 1,
            "max": 214,
            "min": 214,
            "sum": 214
          },
          "bytes-sent": {
            "buckets": [
              0,
              0,
              0,
              1
            ],
            "count": 1,
            "max": 271,
            "min": 271,
            "sum": 271
          },
          "connection-aq": {
            "buckets": [
              1,
              0,
              0,
              0
            ],
            "count": 1,
            "max": 0,
            "min": 0,
            "sum": 0
          },
          "latency": {
            "buckets": [
              0,
              0,
              1,
              0
            ],
            "count": 1,
            "max": 6,
            "min": 6,
            "sum": 6
          },
          "parsing": {
            "buckets": [
              0,
              0,
              0,
              1
            ],
            "count": 1,
            "max": 491,
            "min": 491,
            "sum": 491
          }
        },
        "Put": {
          "bytes-received": {
            "buckets": [
              0,
              0,
              0,
              8
            ],
            "count": 8,
            "max": 30,
            "min": 30,
            "sum": 240
          },
          "bytes-sent": {
            "buckets": [
              0,
              0,
              0,
              8
            ],
            "count": 8,
            "max": 100,
            "min": 100,
            "sum": 800
          },
          "connection-aq": {
            "buckets": [
              8,
              0,
              0,
              0
            ],
            "count": 8,
            "max": 1,
            "min": 0,
            "sum": 1
          },
          "latency": {
            "buckets": [
              0,
              0,
              3,
              5
            ],
            "count": 8,
            "max": 15,
            "min": 5,
            "sum": 64
          },
          "parsing": {
            "buckets": [
              0,
              0,
              0,
              8
            ],
            "count": 8,
            "max": 1263,
            "min": 291,
            "sum": 3575
          }
        }
      }
    },
    "detailed-resultcode-counts": {
      "test": {
        "BatchRead": {
          "OK": 24
        },
        "BatchWrite": {
          "OK": 8
        },
        "Put": {
          "OK": 8
        }
      }
    },
    "exceeded-max-retries": 0,
    "exceeded-total-timeout": 0,
    "exists-metrics": {
      "buckets": [
        0,
        0,
        0,
        0
      ],
      "count": 0,
      "max": 0,
      "min": 0,
      "sum": 0
    },
    "get-header-metrics": {
      "buckets": [
        0,
        0,
        0,
        0
      ],
      "count": 0,
      "max": 0,
      "min": 0,
      "sum": 0
    },
    "get-metrics": {
      "buckets": [
        0,
        0,
        0,
        0
      ],
      "count": 0,
      "max": 0,
      "min": 0,
      "sum": 0
    },
    "labels": [
      {
        "app-id": "",
        "cluster": "",
        "host": "127.0.0.1:3100",
        "node": "BB979E256BB9B1A"
      }
    ],
    "node-added-count": 1,
    "node-removed-count": 0,
    "open-connections": 100,
    "operate-metrics": {
      "buckets": [
        0,
        0,
        0,
        0
      ],
      "count": 0,
      "max": 0,
      "min": 0,
      "sum": 0
    },
    "partition-map-updates": 1,
    "put-metrics": {
      "buckets": [
        0,
        0,
        0,
        8
      ],
      "count": 8,
      "max": 1358,
      "min": 302,
      "sum": 3743
    },
    "query-metrics": {
      "buckets": [
        0,
        0,
        0,
        0
      ],
      "count": 0,
      "max": 0,
      "min": 0,
      "sum": 0
    },
    "scan-metrics": {
      "buckets": [
        0,
        0,
        0,
        0
      ],
      "count": 0,
      "max": 0,
      "min": 0,
      "sum": 0
    },
    "tends-failed": 0,
    "tends-successful": 2,
    "tends-total": 2,
    "transaction-error-count": 0,
    "transaction-retry-count": 0,
    "udf-metrics": {
      "buckets": [
        0,
        0,
        0,
        0
      ],
      "count": 0,
      "max": 0,
      "min": 0,
      "sum": 0
    }
  },
  "open-connections": 100,
  "total-nodes": 1
}
```

mirzakaracic committed:
CLIENT-3365 Metrics changes

khaf commented on base_read_command.go:
```diff
@@ -98,3 +98,10 @@ func (cmd *baseReadCommand) Execute() Error {
 	return ttGet
 }
+
+func (cmd *baseReadCommand) getNamespace() *map[string]uint64 {
```
This method should probably be broken into two: One that returns a string for single namespace commands, and one that returns multiple namespaces. The current implementation imposes too much cost on single namespace commands.

mirzakaracic replied:
Have split getNamespace into multiple methods. Depending on the command we will either send single namespace or set of namespaces.

khaf commented on client_policy.go:
```diff
@@ -161,6 +161,8 @@ type ClientPolicy struct {
 	// Peers nodes for the cluster are not discovered and seed nodes are
 	// retained despite connection failures.
 	SeedOnlyCluster bool // = false
+
```
Need a documentation comment here.

khaf commented on cluster.go:
```diff
@@ -984,9 +986,6 @@ func (clstr *Cluster) MetricsEnabled() bool {
 // If the parameters for the histogram in the policy are the different from the one already
 // on the cluster, the metrics will be reset.
 func (clstr *Cluster) EnableMetrics(policy *MetricsPolicy) {
-	if policy == nil {
```
Why was the default metric policy allocation removed?

khaf commented on cluster.go:
```diff
@@ -1004,6 +1003,49 @@ func (clstr *Cluster) EnableMetrics(policy *MetricsPolicy) {
+		// Users are allowed to override app-id if they want to. Default is the user name.
+		// Users need to set the application id int the client policy.
+		entries["app-id"] = func() string {
+			if clstr.clientPolicy.ApplicationId != "" {
```
Why is a closure being used here?

mirzakaracic replied:
Ternary assignment. This is the equivalent of java 
`var entry = expression ? value : differentValue;`

golang escaping analysis should be able to optimize and inline this however to make sure I have moved this outside closure to avoid extra allocations.

khaf commented on command.go:
```diff
@@ -3690,6 +3723,8 @@ func (cmd *baseCommand) executeAt(ifc command, policy *BasePolicy, deadline time
 			continue
 		}
 
+		metricsEnabled := cmd.node.cluster.metricsEnabled.Load()
```
No need to do a sync read inside of the loop. Assigning it once outside the execution is good enough IMO.

mirzakaracic replied:
I agree. However with dynamic configuration, upcoming PR, users will be able to toggle metrics while the application is running. We should leave this in place. Or perhaps consider not to get the value every iteration but every n cycles? We could expose a variable in metrics policy that could control frequency check.

khaf commented on command.go:
```diff
@@ -3768,7 +3812,17 @@ func (cmd *baseCommand) executeAt(ifc command, policy *BasePolicy, deadline time
+		if metricsEnabled {
+			start := time.Now()
+			dataSent := len(cmd.dataBuffer[:cmd.dataOffset])
+			dataReceived, err = cmd.conn.Write(cmd.dataBuffer[:cmd.dataOffset])
```
`dataReceived` from a `Write`? It returns the same number as `dataSent` unless there was a transmission issue.

mirzakaracic replied:
It's the amount of data put on wire not how much was actually received from the server.

khaf commented on command.go:
```diff
@@ -3790,7 +3844,15 @@ func (cmd *baseCommand) executeAt(ifc command, policy *BasePolicy, deadline time
-		err = ifc.parseResult(ifc, cmd.conn)
+		if metricsEnabled {
+			start := time.Now()
+			err = ifc.parseResult(ifc, cmd.conn)
```
Do we not have the total amount of bytes received?

mirzakaracic replied:
The on new parseResult we  have receiveSize. We could keep that information in parseResult and return parseResult referrence when we call parseResult(...). Something like

```
pr, err = ifc.parseResult(ifc, cmd.conn)
```

Is this falling in line what you thought?

khaf commented on command.go:
```diff
@@ -3965,3 +4028,72 @@ func (cmd *baseCommand) parseVersion(fieldCount int) *uint64 {
 }
+
+// applyDetailedMetricsParsing updates the detailed metrics for parsing time.
+func (cmd *baseCommand) applyDetailedMetricsParsing(ifc command, startTime time.Time) {
```
There are too many allocations here. All should be eliminated, unless absolutely necessary.

khaf commented on command.go:
```diff
@@ -3965,3 +4028,72 @@ func (cmd *baseCommand) parseVersion(fieldCount int) *uint64 {
+}
+
+// applyDetailedMetricsDataSizeAndLatency updates the detailed metrics for bytes sent and transmission time.
+func (cmd *baseCommand) applyDetailedMetricsDataSizeAndLatency(ifc command, bytesSent int, bytesReceived int, startTime time.Time) {
```
This function name does not match what it is used for. It is used when a command is sent to the server.

khaf commented on metadata_policy.go:
```diff
@@ -0,0 +1,19 @@
+	Labels *[]map[string]string
+}
+
+func NewLabels(pairs ...map[string]string) *Labels {
```
Needs documentation. Why are pairs variadic?

mirzakaracic replied:
We can get list of key value pairs or single map with multiple key values.

khaf commented on metrics_policy.go:
```diff
@@ -50,12 +50,37 @@ type MetricsPolicy struct {
 	LatencyBase int //= 2;
+
+	// Labels                  NewMetadataPolicy(map[string]string{"user":policy.User}, map[string]string{"cluster-name": policy.ClusterName}), }
+	Labels *Labels
```
Needs documentation.

khaf commented on metrics_policy.go:
```diff
@@ -50,12 +50,37 @@ type MetricsPolicy struct {
+	}
+}
+
+func DefaultMetricsPolicyWithLabels(pairs ...map[string]string) *MetricsPolicy {
```
Needs documentation.

khaf commented on metrics_policy.go:
```diff
@@ -50,12 +50,37 @@ type MetricsPolicy struct {
+	return &mp
+}
+
+func CopyMetricsPolicyWithLabels(mp *MetricsPolicy, pairs ...map[string]string) *MetricsPolicy {
```
Needs documentation. It is also not clear why this exported function needed.

mirzakaracic replied:
It is not. I have removed it.

khaf commented on node_stats.go:
```diff
@@ -396,6 +526,190 @@ func (ns *nodeStats) UnmarshalJSON(data []byte) error {
 }
+
+// Clone returns a deep copy of the nodeStats DetailedMetrics object
+func (ns *nodeStats) cloneDetailedMetrics() *amap.Map[string, *amap.Map[commandType, *commandMetric]] {
```
I suspect the closures inside this and the following functions are very expensive, but since they are not in the hot path and only by demand every few seconds, I think we can leave them be. Although benching them is a good idea to make sure they don't take more than a few milliseconds.

mirzakaracic replied:
It is expensive but we are not getting close to milisconds per operation 

BenchmarkCloneDetailedMetrics-14    	 1208846	      1016 ns/op	    3416 B/op	      24 allocs/op
BenchmarkCloneAndResetDetailedResultCodeCounts-14    	 2559760	       466.9 ns/op	     888 B/op	      14 allocs/op

khaf commented on types/histogram/sync_histogram.go:
```diff
@@ -138,9 +138,14 @@ func (h *SyncHistogram[T]) CloneAndReset() *SyncHistogram[T] {
 	h.l.Lock()
-	other.l.RLock()
+
+	// It is possible to try to get lock on the same histogram in the same goroutine
```
We never merge the same histogram on itself. This change is not necessary and defer is expensive. You also ignore the case when the lock acquisition fails on `other`.

khaf reviewed (changes requested):
Looks mostly good. There are a few issues that need addressing:

- [x] Some closures are unnecessary and need removal.
- [x] Some closures in the hot path look to be costly both in terms of GC pressure and performance and need evaluation.
- [x] The amount of data received from the server is not recorded in the stats.
- [x] Histogram merge seems to have been broken.

mirzakaracic committed:
Addressing most comments

mirzakaracic committed:
Removed all use of closures

mirzakaracic committed:
Minor cleanup

khaf commented on batch_command_operate.go:
```diff
@@ -301,3 +307,24 @@ func (cmd *batchCommandOperate) commandType() commandType {
+
+func (cmd *batchCommandOperate) getNamespaces() *map[string]uint64 {
+	response := make(map[string]uint64, len(cmd.records))
+	for _, br := range cmd.records {
```
There is a `BatchRecordIfc.key()` method that will obviate the switch here.

khaf commented:
We had a good discussion with Mirza, and made a few improvements, the most important of which is that now the client can reliably return the amount of data received from the server. We also reduced the number of allocations where possible.

mirzakaracic commented:
I think we are good on the total bytes received. I have have performed the following operations

```
size := 8

util.WriteRecords(cli, keyPrefix, binName, valuePrefix, size)
util.BatchExists(cli, keyPrefix, size)
util.BatchReads(cli, keyPrefix, binName, size)
util.BatchReadHeaders(cli, keyPrefix, size)
util.BatchDelete(cli, keyPrefix, size)
```

The output for the commands ran above in the same sequence are 

```
2025/05/02 12:30:57 Starting the cluster tend goroutine...
Put: ns=test set=demo key=batchkey0 bin=batchbin value=batchvalue0
2025/05/02 12:30:57 Node BB979E256BB9B1A [::1]:3100: Received 30 bytes, command type: Put
Put: ns=test set=demo key=batchkey1 bin=batchbin value=batchvalue1
2025/05/02 12:30:57 Node BB979E256BB9B1A [::1]:3100: Received 30 bytes, command type: Put
Put: ns=test set=demo key=batchkey2 bin=batchbin value=batchvalue2
2025/05/02 12:30:57 Node BB979E256BB9B1A [::1]:3100: Received 30 bytes, command type: Put
Put: ns=test set=demo key=batchkey3 bin=batchbin value=batchvalue3
2025/05/02 12:30:57 Node BB979E256BB9B1A [::1]:3100: Received 30 bytes, command type: Put
Put: ns=test set=demo key=batchkey4 bin=batchbin value=batchvalue4
2025/05/02 12:30:57 Node BB979E256BB9B1A [::1]:3100: Received 30 bytes, command type: Put
Put: ns=test set=demo key=batchkey5 bin=batchbin value=batchvalue5
2025/05/02 12:30:57 Node BB979E256BB9B1A [::1]:3100: Received 30 bytes, command type: Put
Put: ns=test set=demo key=batchkey6 bin=batchbin value=batchvalue6
2025/05/02 12:30:57 Node BB979E256BB9B1A [::1]:3100: Received 30 bytes, command type: Put
Put: ns=test set=demo key=batchkey7 bin=batchbin value=batchvalue7
2025/05/02 12:30:57 Node BB979E256BB9B1A [::1]:3100: Received 30 bytes, command type: Put
BatchExists creating 8 keys
2025/05/02 12:30:57 Node BB979E256BB9B1A [::1]:3100: Received 214 bytes, command type: BatchExists
2025/05/02 12:30:57 Record: ns=test set=demo key=batchkey0 exists=true
2025/05/02 12:30:57 Record: ns=test set=demo key=batchkey1 exists=true
2025/05/02 12:30:57 Record: ns=test set=demo key=batchkey2 exists=true
2025/05/02 12:30:57 Record: ns=test set=demo key=batchkey3 exists=true
2025/05/02 12:30:57 Record: ns=test set=demo key=batchkey4 exists=true
2025/05/02 12:30:57 Record: ns=test set=demo key=batchkey5 exists=true
2025/05/02 12:30:57 Record: ns=test set=demo key=batchkey6 exists=true
2025/05/02 12:30:57 Record: ns=test set=demo key=batchkey7 exists=true
BatchRead creating 8 keys
2025/05/02 12:30:57 Node BB979E256BB9B1A [::1]:3100: Received 430 bytes, command type: BatchRead
Record: ns=test set=demo key=batchkey0 bin=batchbin value=batchvalue0
Record: ns=test set=demo key=batchkey1 bin=batchbin value=batchvalue1
Record: ns=test set=demo key=batchkey2 bin=batchbin value=batchvalue2
Record: ns=test set=demo key=batchkey3 bin=batchbin value=batchvalue3
Record: ns=test set=demo key=batchkey4 bin=batchbin value=batchvalue4
Record: ns=test set=demo key=batchkey5 bin=batchbin value=batchvalue5
Record: ns=test set=demo key=batchkey6 bin=batchbin value=batchvalue6
Record: ns=test set=demo key=batchkey7 bin=batchbin value=batchvalue7
BatchReadHeaders creating 8 keys
2025/05/02 12:30:57 Node BB979E256BB9B1A [::1]:3100: Received 214 bytes, command type: BatchReadHeader 
Record: ns=test set=demo key=batchkey0 generation=1 expiration=4294967295
Record: ns=test set=demo key=batchkey1 generation=1 expiration=4294967295
Record: ns=test set=demo key=batchkey2 generation=1 expiration=4294967295
Record: ns=test set=demo key=batchkey3 generation=1 expiration=4294967295
Record: ns=test set=demo key=batchkey4 generation=1 expiration=4294967295
Record: ns=test set=demo key=batchkey5 generation=1 expiration=4294967295
Record: ns=test set=demo key=batchkey6 generation=1 expiration=4294967295
Record: ns=test set=demo key=batchkey7 generation=1 expiration=4294967295
BatchReadHeaders creating 8 keys
2025/05/02 12:30:57 Node BB979E256BB9B1A [::1]:3100: Received 214 bytes, command type: BatchDelete
Record: ns=test set=demo key=batchkey0 resultCode=0
Record: ns=test set=demo key=batchkey1 resultCode=0
Record: ns=test set=demo key=batchkey2 resultCode=0
Record: ns=test set=demo key=batchkey3 resultCode=0
Record: ns=test set=demo key=batchkey4 resultCode=0
Record: ns=test set=demo key=batchkey5 resultCode=0
Record: ns=test set=demo key=batchkey6 resultCode=0
Record: ns=test set=demo key=batchkey7 resultCode=0
```

mirzakaracic committed:
Discussed changes

khaf committed:
Convert command.getNamespaces to return a Iter.Seq2 instead of a map

The map will forever require heap allocations. The Iterator on the other hand can be optimized away to NOT require allocations. At the moment Go's compiler is not that sophisticated and still allocates memory for the iterator, but I expect that to be resolved in the future. Having said that, Iterator still performs significantly better and allocates less memory even today.

mirzakaracic commented:
So the reason we dont have the data corruption issue with maps is because we are creating a copy. This is also the overhead. We most likely could fix the issue when using iterator with locks maybe, I have not really thought this through, but since number of namespaces is not unbound I don't think it make sense to use a generator here.

There are swissmaps coming to golang in 1.24 as default implementation for map.

mirzakaracic committed:
Fix for misleading data races.

khaf reviewed (approved):
Looks good.
